readthedocs/core/views/hooks.py

```diff
@@ -88,3 +91,92 @@ def sync_versions(project):
     except Exception:
         log.exception('Unknown sync versions exception')
     return None
+
+
+def get_or_create_external_version(project, identifier, verbose_name):
+    """
+    Get or create external versions using `identifier` and `verbose_name`.
+
+    if external version does not exist create an external version
+
+    :param project: Project instance
+    :param identifier: Commit Hash
+    :param verbose_name: pull/merge request number
+    :returns:  External version.
+    :rtype: Version
+    """
+    external_version = project.versions(
+        manager=EXTERNAL
+    ).filter(verbose_name=verbose_name).first()
+
+    if external_version:
+        # identifier will change if there is a new commit to the Pull/Merge Request
+        if external_version.identifier != identifier:
+            external_version.identifier = identifier
+            external_version.save()
+    else:
+        # create an external version if the version does not exist.
+        created_external_version = Version.objects.create(
+            project=project,
+            type=EXTERNAL,
+            identifier=identifier,
+            verbose_name=verbose_name,
+            active=True
+        )
+        log.info(
+            '(Create External Version) Added Version: [%s] ', ' '.join(
+                created_external_version.slug
+            )
+        )
+        return created_external_version
+    return external_version
+
+
+def delete_external_version(project, identifier, verbose_name):
+    """
+    Delete external versions using `identifier` and `verbose_name`.
+
+    if external version does not exist then returns `None`.
+
+    :param project: Project instance
+    :param identifier: Commit Hash
+    :param verbose_name: pull/merge request number
+    :returns:  verbose_name (pull/merge request number).
+    :rtype: str
+    """
+    external_version = project.versions(manager=EXTERNAL).filter(
```

Comment:
I did not call delete from the queryset because then we wont be able to get an exact version that was deleted and log that also send it with the response data :)